### PR TITLE
Optional discount date #2763253 

### PR DESCRIPTION
**Pull request by mglaman**
… not set

### CODE REVIEW

File: modules/promotion/src/Entity/Promotion.php [hunk 1]
```diff
@@ -293,14 +310,26 @@ public static function baseFieldDefinitions(EntityTypeInterface $entity_type) {
...
-      ->setRequired(TRUE)
+      ->setRequired(FALSE)
       ->setSetting('datetime_type', 'date')
       ->setDefaultValueCallback('Drupal\commerce_promotion\Entity\Promotion::getDefaultEndDate')
       ->setDisplayOptions('form', [
         'type' => 'datetime_default',
         'weight' => 4,
       ]);
 
+    $fields['has_end_date'] = BaseFieldDefinition::create('boolean')
```
Comment (bojanz) [inline modules/promotion/src/Entity/Promotion.php:321]: We don't need a field for this, we can infer it from whether the end_date field is empty.

File: modules/promotion/src/Entity/Promotion.php [hunk 2]
```diff
@@ -205,6 +207,21 @@ public function setEndDate(DrupalDateTime $end_date) {
...
    * {@inheritdoc}
    */
+  public function hasEndDate() {
+    return (bool) $this->getEntityKey('has_end_date');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setHasEndDate($hasEndDate) {
```
Comment (bojanz) [inline modules/promotion/src/Entity/Promotion.php:217]: No need for a setter. hasEndDate can operate on the end_date field directly.

File: modules/promotion/src/Entity/Promotion.php [hunk 3]
```diff
@@ -205,6 +207,21 @@ public function setEndDate(DrupalDateTime $end_date) {
   /**
    * {@inheritdoc}
    */
+  public function hasEndDate() {
+    return (bool) $this->getEntityKey('has_end_date');
```
Comment (bojanz) [inline modules/promotion/src/Entity/Promotion.php:211]: getEntityKey() is not used for accessing ordinary fields.

File: modules/promotion/src/Entity/Promotion.php [hunk 4]
```diff
@@ -342,3 +371,5 @@ public static function getDefaultEndDate() {
   }
 
 }
+
+
```
Comment (bojanz) [inline modules/promotion/src/Entity/Promotion.php:375]: Extra newlines.

File: modules/promotion/src/Entity/Promotion.php [hunk 5]
```diff
@@ -26,8 +27,9 @@
  *     "list_builder" = "Drupal\commerce_promotion\PromotionListBuilder",
  *     "views_data" = "Drupal\views\EntityViewsData",
  *     "form" = {
- *       "add" = "Drupal\Core\Entity\ContentEntityForm",
- *       "edit" = "Drupal\Core\Entity\ContentEntityForm",
+ *       "default" = "Drupal\commerce_promotion\Form\PromotionForm",
+ *       "add" = "Drupal\Core\Entity\Form\PromotionForm",
```
Comment (mglaman) [inline modules/promotion/src/Entity/Promotion.php:31]: Wrong FQCN

File: modules/promotion/src/Entity/Promotion.php [hunk 6]
```diff
@@ -342,3 +371,5 @@ public static function getDefaultEndDate() {
   }
 
 }
+
```
Comment (mglaman) [inline modules/promotion/src/Entity/Promotion.php:374]: Extra lines

File: modules/promotion/src/Form/PromotionForm.php [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ */
+class PromotionForm extends ContentEntityForm {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function form(array $form, FormStateInterface $form_state) {
+    $form = parent::form($form, $form_state);
+
+    $form['end_date']['#states'] = [
```
Comment (bojanz) [inline modules/promotion/src/Form/PromotionForm.php:19]: This should be done as a custom end_date widget ("optional_date" or something)

**Comment by mglaman** [comment]
Closing in favor of #441 where I have addressed feedback.

**Pull request closed without merging**